### PR TITLE
Be able to retrieve XSRF token with session cookie

### PR DESCRIPTION
**Pull request by shawnzhu**
For #8441

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

**Review by rchiodo — APPROVED** [review]

**Comment by rchiodo** [comment]
@shawnzhu can you sign the license agreement? We can't submit without it.

**Comment by rchiodo** [comment]
It looks like the unit tests are failing from the code you changed too:

```
getPasswordConnectionInfo:1 | Cannot read property 'ok' of undefined
getPasswordConnectionInfo allowUnauthorized:1 | Cannot read property 'ok' of undefined
getPasswordConnectionInfo failure:1 | Cannot read property 'ok' of undefined
```

**Review by DonJayamanne — APPROVED** [review]

File: src/client/datascience/jupyter/jupyterPasswordConnect.ts [hunk 1]
```diff
@@ -315,13 +320,23 @@ export class JupyterPasswordConnect implements IJupyterPasswordConnect {
...
+        let headers = {
+            Connection: 'keep-alive',
+            Cookie: ''
+        };
+        let tokenUrl = `${url}login?`;
+
+        if (sessionCookie != '') {
+            tokenUrl = `${url}tree`;
+            headers.Cookie = sessionCookie;
+        }
```
Comment (PrimeF) [inline src/client/datascience/jupyter/jupyterPasswordConnect.ts:334]:
```suggestion
        let headers;
        let tokenUrl = `${url}login?`;

        if (sessionCookie != '') {
            tokenUrl = `${url}tree`;
            headers = {
                Connection: 'keep-alive',
                Cookie: sessionCookie,
            };
        } else {
            headers = {
                Connection: 'keep-alive',
            };
        }
```

File: src/client/datascience/jupyter/jupyterPasswordConnect.ts [hunk 2]
```diff
@@ -315,13 +320,23 @@ export class JupyterPasswordConnect implements IJupyterPasswordConnect {
...
 
-        const response = await this.makeRequest(`${url}login?`, {
+        const response = await this.makeRequest(tokenUrl, {
             method: 'get',
             redirect: 'manual',
-            headers: { Connection: 'keep-alive' }
+            headers,
         });
 
         if (response.ok) {
```
Comment (PrimeF) [inline src/client/datascience/jupyter/jupyterPasswordConnect.ts:342]:
```suggestion
        if (response !== undefined && response.ok) {
```

**Review by PrimeF — CHANGES_REQUESTED** [review]
The tests are failing because in case of a missing XSRF token they require the connection to fail. Instead, the Kubeflow logic additionally requests an XSRF token using the session cookie, which leads to an undefined response.

File: src/client/datascience/jupyter/jupyterPasswordConnect.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ export class JupyterPasswordConnect implements IJupyterPasswordConnect {
         @inject(IMultiStepInputFactory) private readonly multiStepFactory: IMultiStepInputFactory,
         @inject(IAsyncDisposableRegistry) private readonly asyncDisposableRegistry: IAsyncDisposableRegistry,
         @inject(IConfigurationService) private readonly configService: IConfigurationService
-    ) {}
+    ) { }
```
Comment (PrimeF) [inline src/client/datascience/jupyter/jupyterPasswordConnect.ts:28]:
```suggestion
    ) {}
```

File: src/client/datascience/jupyter/jupyterPasswordConnect.ts [hunk 4]
```diff
@@ -223,13 +223,18 @@ export class JupyterPasswordConnect implements IJupyterPasswordConnect {
...
-                xsrfCookie = await this.getXSRFToken(url);
+                xsrfCookie = await this.getXSRFToken(url, '');
 
                 // Then get the session cookie by hitting that same page with the xsrftoken and the password
                 if (xsrfCookie) {
                     const sessionResult = await this.getSessionCookie(url, xsrfCookie, userPassword);
                     sessionCookieName = sessionResult.sessionCookieName;
                     sessionCookieValue = sessionResult.sessionCookieValue;
+                } else { // get xsrf cookie with session cookie
+                    sessionCookieName = "authservice_session";
```
Comment (PrimeF) [inline src/client/datascience/jupyter/jupyterPasswordConnect.ts:234]:
```suggestion
                } else {
                    // get xsrf cookie with session cookie
                    sessionCookieName = 'authservice_session';
```

File: src/client/datascience/jupyter/jupyterPasswordConnect.ts [hunk 5]
```diff
@@ -315,16 +320,30 @@ export class JupyterPasswordConnect implements IJupyterPasswordConnect {
...
+            headers = {
+                Connection: 'keep-alive',
+            };
+        }
+
+        const response = await this.makeRequest(tokenUrl, {
             method: 'get',
             redirect: 'manual',
-            headers: { Connection: 'keep-alive' }
+            headers,
```
Comment (PrimeF) [inline src/client/datascience/jupyter/jupyterPasswordConnect.ts:343]:
```suggestion
            headers
```

File: src/client/datascience/jupyter/jupyterPasswordConnect.ts [hunk 6]
```diff
@@ -315,16 +320,30 @@ export class JupyterPasswordConnect implements IJupyterPasswordConnect {
...
-        const response = await this.makeRequest(`${url}login?`, {
+        if (sessionCookie != '') {
+            tokenUrl = `${url}tree`;
+            headers = {
+                Connection: 'keep-alive',
+                Cookie: sessionCookie,
+            };
+        } else {
+            headers = {
+                Connection: 'keep-alive',
```
Comment (PrimeF) [inline src/client/datascience/jupyter/jupyterPasswordConnect.ts:336]:
```suggestion
                Connection: 'keep-alive'
```

File: src/client/datascience/jupyter/jupyterPasswordConnect.ts [hunk 7]
```diff
@@ -315,16 +320,30 @@ export class JupyterPasswordConnect implements IJupyterPasswordConnect {
...
         let xsrfCookie: string | undefined;
+        let headers;
+        let tokenUrl = `${url}login?`;
 
-        const response = await this.makeRequest(`${url}login?`, {
+        if (sessionCookie != '') {
+            tokenUrl = `${url}tree`;
+            headers = {
+                Connection: 'keep-alive',
+                Cookie: sessionCookie,
```
Comment (PrimeF) [inline src/client/datascience/jupyter/jupyterPasswordConnect.ts:332]:
```suggestion
                Cookie: sessionCookie
```

**Review by PrimeF — CHANGES_REQUESTED** [review]
This should fix the Linter test

**Pull request merged**